incubator/hnc/test/e2e/demo_test.go

```diff
@@ -93,8 +93,91 @@ var _ = Describe("Demo", func() {
 		RunShouldNotContain("my-creds", 2, "kubectl -n", nsService1, "get secrets")
 	})
 
-	It("Should create and delete subnamespaces", func() {
-		// set up initial structure
+	It("Should intergrate hierarchical network policy", func(){
+		MustRun("kubectl create ns", nsOrg)
+		MustRun("kubectl hns create", nsTeamA, "-n", nsOrg)
+		MustRun("kubectl hns create", nsTeamB, "-n", nsOrg)
+		MustRun("kubectl hns create", nsService1, "-n", nsTeamA)
+		MustRun("kubectl hns create", nsService2, "-n", nsTeamA)
+		// create a web service s2 in namespace service-2, and a client pod client-s1 in namespace service-1 that can access this web service
+		MustRun("kubectl run s2 -n", nsService2, "--image=nginx --restart=Never --expose --port 80")
+		// at least 20 seconds is needed here from experiments 
+		RunShouldContain("Welcome to nginx!", 20, 
+			"kubectl run client -n", nsService1, "-i --image=alpine --restart=Never --rm -- sh -c", "\"wget -qO- --timeout 2 http://s2.service-2\"")
+		RunShouldContain("Welcome to nginx!", 10, 
+			"kubectl run client -n", nsTeamA, "-i --image=alpine --restart=Never --rm -- sh -c", "\"wget -qO- --timeout 2 http://s2.service-2\"")
+		RunShouldContain("Welcome to nginx!", 10, 
+			"kubectl run client -n", nsTeamB, "-i --image=alpine --restart=Never --rm -- sh -c", "\"wget -qO- --timeout 2 http://s2.service-2\"")
+
+		// create a default network policy that blocks any ingress from other namespaces 
+		policy := "" +
+			"kind: NetworkPolicy" + "\n" +
+			"apiVersion: networking.k8s.io/v1" + "\n" +
+			"metadata:" + "\n" +
+			"  name: deny-from-other-namespaces" + "\n" +
+			"  namespace: acme-org" + "\n" +
+			"spec:" + "\n" +
+			"  podSelector:" + "\n" +
+			"    matchLabels:" + "\n" +
+			"  ingress:" + "\n" +
+			"  - from:" + "\n" +
+			"    - podSelector: {}"
+		filename := "temp.yaml"
+		WriteToFile(policy, filename)
+
+		MustRun("kubectl apply -f", filename)
+		// ensure this policy can be propagated to its descendants
+		MustRun("kubectl hns config set-type --apiVersion networking.k8s.io/v1 --kind NetworkPolicy propagate")
+		expected := []string {
+			nsOrg 	+ "    deny-from-other-namespaces",
+			nsService1 + "   deny-from-other-namespaces",
+			nsService2 + "   deny-from-other-namespaces",
+			nsTeamA + "      deny-from-other-namespaces",
+			nsTeamB + "      deny-from-other-namespaces",
+		}
+		RunShouldContainMultiple(expected, 2, "kubectl get netpol --all-namespaces")
```

Comment:
The constant number of strings seems quite brittle. Maybe better to iterate through the namespaces one-by-one and call `RunShouldContain` in each namespace individually.

Reply:
done